api docs: Add an API changelog page.

timabbott opened the pull request:
I imagine this can be improved in various ways, but I've initialized
this with all the **Changes** entries recorded in either zulip.yaml or
the rest of the API documentation, and I expect we'll be able to
iterate on this effectively.

![image](https://user-images.githubusercontent.com/2746074/80559724-5c4d7780-8993-11ea-842a-b0445a992661.png)

gnprice commented:
Thanks for collecting these!

Did you see my reply to you at https://github.com/zulip/zulip/pull/14658#issuecomment-617411240 ?

---

> @gnprice what are your thoughts on whether we want the API documentation to be just a long bullet on the `/server_settings` page, or e.g. a separate "changelog" page entirely? If it's the latter, we should probably take the opportunity to initialize that page; if we're happy as-is, then we don't need to do that.

I think the place I most want it is inline where the respective features appear in the API docs. At least for myself and mobile, I don't think there's a strong need to have a changelog collected in one place. (So long as we continue to have a protocol where you mention new changes that we should be aware of, anyway -- I think the issues you've been filing with label "api migration" have been working pretty well, and we can continue to iterate as needed.)

For that, I think the initialization that's useful to do is, in the handful of places that now mention things new in Zulip 2.2, to edit that to like "Zulip 2.2 (feature level 1)", provided that's accurate. A grep like `git grep -C8 "Zulip 2.2" templates/zerver/api/` finds two, in get-messages and get-all-users.

---

I could see this central list being helpful in the workflow of keeping these changes documented -- it can provide a handy place for you to compare to your mental tracking of what's changed, as well as to compare to other forms of changelog. But I think from the perspective of writing client code, I'd still like to have it mentioned inline where the respective features appear in the API docs.

gnprice commented on templates/zerver/api/changelog.md:
```diff
@@ -0,0 +1,78 @@
+
+* [`GET /users`](/api/get-all-users): Added `include_custom_profile_fields`
+  to request custom profile field data.
+* [`GET /users/me`](/api/get-profile): Added `avatar_url` field,
+  containing the user's avatar URL, to the response.
+* [`GET /users/me/subscriptions`](/api/get-subscribed-streams): Added
+  `include_subscribers` parameter controlling whether data on the
+  other subscribers is included.  Previous behavior was to always send
+  subscriber data.
+* [`GET /server_settings`](/api/server-settings): Added
+  `external_authentication_methods` structure, used to display login
+  buttons nicely in the mobile apps.
```
The `/users/me` and `/server_settings` changes aren't mentioned in those docs pages -- it'd be good to add them.

(Hmm, also most of `/users/me` isn't really documented! Most of the properties could be covered well by a reference to get-user or get-all-users. But I'm not acutally sure what to make of `client_id`, and I haven't gone through the whole list to find what else there is.)

timabbott replied:
Agreed.

We currently don't have a standard system for responses; so part of the reason I'm adding mentions of things like that here is as reminders to document them when we do add the responses data.

Part of our master OpenAPI docs plan is to make the "responses" section for API endpoints work like the "parameters" section, generated from the OpenAPI data and validated by our test suite.

gnprice commented on templates/zerver/api/changelog.md:
```diff
@@ -0,0 +1,78 @@
+* [`POST /users/me/subscriptions`](/api/add-subscriptions): Replaced
+  `is_announcement_only` boolean with `stream_post_policy` enum for
+  specifying who can post to a stream.
+* [`PATCH /streams/{stream_id}`](/api/update-stream): Replaced
+  `is_announcement_only` boolean with `stream_post_policy` enum for
+  specifying who can post to a stream.
+* [`GET /streams/{stream_id}`](/api/update-stream): Replaced
+  `is_announcement_only` boolean with `stream_post_policy` enum for
+  specifying who can post to a stream.
```
These two link to the same place, which says `PATCH`. Is the `GET` something else?

timabbott replied:
Good catch.

gnprice commented on templates/zerver/api/changelog.md:
```diff
@@ -0,0 +1,78 @@
+
+**Feature level 2**:
+
+* [`POST /messages/{message_id}/reactions`](/api/add-emoji-reaction):
+  The `emoji_type` parameter is optional; the server will guess the
+  emoji_type if it is not specified (checking custom emoji, then
+  unicode emoji for any with the provided name).
```
This change isn't documented on the endpoint's page.

timabbott replied:
Yeah, adding that in a next commit.

timabbott commented:
Pushed an update to this, that adds a bunch of extra things to the Changelog section for Zulip 2.1 (E.g. new organization settings that would appear in responses).  

> I think the place I most want it is inline where the respective features appear in the API docs.

I've added `(Zulip feature level 1)` notices everywhere 2.2 is mentioned that isn't "deprecated" type notices (where the feature level doesn't really apply).  Note that we just moved the description content into `zerver/openapi/zulip.yaml`; the actual `templates/zerver/api` pages are increasingly becoming shells for the REST endpoints.  So remember to include `zerver/openapi/zulip.yaml` in any greps for content.

I'll add a reference to https://github.com/zulip/zulip/issues/14100 for the current near-term master OpenAPI TODO list, which for the moment focuses on making our OpenAPI docs standards-compliant and easy to work with.  One of the most important pieces if making the `response` data have a system like the one for parameters that makes it maintainable so that we can make it actually useful.

@akashaviator @shubhamgupta2956 @orientor FYI.

timabbott commented:
Merging this so that others can continue iterating on top of it.
